Course import on site creation

melvinsoft opened the pull request:
This part of the trial revamp objective of Q1. I reworked Filip's old function. 

The PR introduces the following new settings:
* `IMPORT_TAHOE_DEFAULT_COURSE_ON_SITE_CREATION` this goes inside `FEATURES` dict, it's a boolean to activate the course import or not.
* `TAHOE_DEFAULT_COURSE_NAME`: The name of the course.
* `TAHOE_DEFAULT_COURSE_GITHUB_ORG`: The org in Github.
* `TAHOE_DEFAULT_COURSE_GITHUB_NAME`: The course name in Github (as it goes in the URL).
* `TAHOE_DEFAULT_COURSE_VERSION`: The version or release of the course.

The new course import function is triggered in a celery task when the site is being created, pulls the course from github, imports it and add the current user as course admin.

thraxil commented on openedx/core/djangoapps/appsembler/sites/tasks.py:
```diff
@@ -1,58 +1,106 @@
+        course_version = settings.TAHOE_DEFAULT_COURSE_VERSION
+
+        # create a dir in tmp to download the default course
+        dir = os.path.join("/", "tmp", str(uuid.uuid4()))
```
Could we use `tempfile.mkdtemp()` instead? Wouldn't have to do the uuid stuff or check `os.path.exists()`.

If the `backports` library is available (not sure offhand whether edx already installs it), `tempfile.TemporaryDirectory()` would be even better:

```
from backports import tempfile
with tempfile.TemporaryDirectory() as dir:
    ...
```

it would automatically delete the directory when done so the `os.remove()` and `shutil.rmtree()` stuff below could also be removed.

OmarIthawi replied:
What Anders said.

melvinsoft replied:
@thraxil @OmarIthawi Change implemented.

OmarIthawi commented on openedx/core/djangoapps/appsembler/sites/serializers.py:
```diff
@@ -135,11 +135,9 @@ def create(self, validated_data):
-            destination_course_locator = CourseLocator(organization.name, 'My first course', '2017')
-            clone_course.apply_async(
-                (unicode(source_course_locator), unicode(destination_course_locator), user.id),
+        if settings.FEATURES["IMPORT_TAHOE_DEFAULT_COURSE_ON_SITE_CREATION"]:
```
```suggestion
        if settings.FEATURES["APPSEMBLER_IMPORT_DEFAULT_COURSE_ON_SITE_CREATION"]:
```

OmarIthawi commented on openedx/core/djangoapps/appsembler/sites/tasks.py:
```diff
@@ -1,58 +1,106 @@
+            course_github_org,
+            course_github_name,
+            course_version
+        )
```
```suggestion
```

melvinsoft replied:
@OmarIthawi Sorry, I accepted your suggestion by mistake and then I re applied the change, I don't get why to remove this code, is how we construct the URL.

OmarIthawi commented on openedx/core/djangoapps/appsembler/sites/tasks.py:
```diff
@@ -1,58 +1,106 @@
+        course_name = settings.TAHOE_DEFAULT_COURSE_NAME
+        course_github_org = settings.TAHOE_DEFAULT_COURSE_GITHUB_ORG
+        course_github_name = settings.TAHOE_DEFAULT_COURSE_GITHUB_NAME
+        course_version = settings.TAHOE_DEFAULT_COURSE_VERSION
```
```suggestion
        course_download_url = settings.TAHOE_DEFAULT_COURSE_DOWNLOAD_URL
```

melvinsoft replied:
@OmarIthawi I'm sorry, but I don't get this suggested change.

OmarIthawi commented on openedx/core/djangoapps/appsembler/settings/settings/aws_lms.py:
```diff
@@ -101,3 +101,8 @@ def plugin_settings(settings):
+    settings.TAHOE_DEFAULT_COURSE_NAME = settings.ENV_TOKENS.get('TAHOE_DEFAULT_COURSE_NAME', '')
+    settings.TAHOE_DEFAULT_COURSE_GITHUB_ORG = settings.ENV_TOKENS.get('TAHOE_DEFAULT_COURSE_GITHUB_ORG', '')
+    settings.TAHOE_DEFAULT_COURSE_GITHUB_NAME = settings.ENV_TOKENS.get('TAHOE_DEFAULT_COURSE_GITHUB_NAME', '')
+    settings.TAHOE_DEFAULT_COURSE_VERSION = settings.ENV_TOKENS.get('TAHOE_DEFAULT_COURSE_VERSION', '')
```
```suggestion
    settings.TAHOE_DEFAULT_COURSE_DOWNLOAD_URL = settings.ENV_TOKENS.get('TAHOE_DEFAULT_COURSE_DOWNLOAD_URL', '')
```

melvinsoft replied:
@OmarIthawi Sorry, I don't get this change.

OmarIthawi replied:
I'd like the feature to be decoupled from GitHub so we could use S3 if we want to.

OmarIthawi commented on openedx/core/djangoapps/appsembler/sites/tasks.py:
```diff
@@ -1,58 +1,106 @@
+        # download the course in tar.gz format from github in /tmp using a
+        # random file name
+        r = requests.get(course_download_url)
+        course_file_name = str(uuid.uuid4()) + ".zip"
```
`zip`?

melvinsoft replied:
Done.

OmarIthawi commented on openedx/core/djangoapps/appsembler/sites/tasks.py:
```diff
@@ -1,58 +1,106 @@
+        file_path = os.path.join("/", "tmp", course_file_name)
+        with open(file_path, 'wb') as fd:
+            for chunk in r.iter_content(chunk_size=128):
+                fd.write(chunk)
```
Please consider the [`copyfileobj` shortcut](https://stackoverflow.com/a/39217788/161278). I really hate for-loops for data transfer, it makes me think that I'm writing a C program:

```suggestion
import requests
import shutil

def download_file(url):
    local_filename = url.split('/')[-1]
    with requests.get(url, stream=True) as r:
        with open(local_filename, 'wb') as f:
            shutil.copyfileobj(r.raw, f)

    return local_filename
```

melvinsoft replied:
@OmarIthawi It looks like C because it's old, we're using an old version of `requests` in edx that doesn't allow streams, so I tried it but it was failing, please check this: https://github.com/psf/requests/issues/4157 it was going to be released in 2017, and we're using a version from 2015.

OmarIthawi commented on openedx/core/djangoapps/appsembler/sites/tasks.py:
```diff
@@ -61,18 +109,14 @@ def clone_course(source_course_key_string, destination_course_key_string, user_i
         try:
             # cleanup any remnants of the course
-            modulestore().delete_course(destination_course_key, user_id)
+            modulestore().delete_course(course_target_id, organization.users.first())
```
The task should get the final `user_id` in my opinion instead of trying to calculate it.

```suggestion
            modulestore().delete_course(course_target_id, user_id)
```

melvinsoft replied:
@OmarIthawi It makes more sense in this way for, the user needs to belong to the org, encapsulating the logic in the function protects someone adding a user from another org to the course, we have had that problem in the past, of users from one org, belonging to courses in other org.

OmarIthawi commented on openedx/core/djangoapps/appsembler/sites/serializers.py:
```diff
@@ -135,11 +135,9 @@ def create(self, validated_data):
+        if settings.FEATURES["IMPORT_TAHOE_DEFAULT_COURSE_ON_SITE_CREATION"]:
+            import_course_on_site_creation.apply_async(
+                organization,
                 queue="edx.core.cms.high"
```
> `queue="edx.core.cms.high"`

It runs in CMS, right?

melvinsoft replied:
Yes, it runs in the CMS queue.

OmarIthawi replied:
Amazing!

melvinsoft commented:
@thraxil @OmarIthawi Thank you for your reviews. 

@OmarIthawi I replied inline or addressed all your comments. 

@thraxil I like the backports idea, but is not available in the platform, I'd like you know your opinions if it worth to add a new requirements just for this use case.

OmarIthawi commented:
@melvinsoft it's up to you. The backports package sounds like a nice way to do it. Otherwise you could achieve the same with `from tempfile import mkdtemp`. However, you have to do the cleanup yourself `shutil.rmtree`. The `with tempfile.TemporaryDirectory() as dir:` provides a much nicer cleanup and slight guarantees that it would do its best to cleanup in the case of an Exception for example.

OmarIthawi commented:
> @OmarIthawi Sorry, I don't get this change about the GitHub URLs.

@melvinsoft I don't see why GitHub is needed, while we could use just a URL that we can put from anywhere e.g. S3 in case GitHub decided to go down for a day or two.

GitHub releases are meant to be used (in my opinion) for deployment time, but not for production and be up 100% of the time.

On the other hand, something like S3 or just a simple CloudFront cache in front of GitHub releases are by design much more resilient.

So my suggestion was to use a URL that we could configure to use GitHub or anything in the `edx-configs`.

melvinsoft commented:
@OmarIthawi I've implemented the backports change suggested by @thraxil already, I agree, it's way more elegant.

What I don't agree is the other suggestion, it was designed to use github on purpose for several reasons, for example, if we use a bucket, we need to take care of that bucket, if some reason we re-provision or change our infrastructure, someone will need to upload the course, or several version of the course to the new bucket, also this gives more engineering independence to CS team to work, update and version the course, it also assures the course will be always in tar.gz format, has the correct name, correct folder structure, etc. 
I also did a quick search about github releases availability and I didn't find anything that suggest that is less reliable that github itself.

OmarIthawi commented:
> @OmarIthawi I've implemented the backports change suggested by @thraxil already, I agree, it's way more elegant.
> 
> What I don't agree is the other suggestion, it was designed to use github on purpose for several reasons, for example, if we use a bucket, we need to take care of that bucket, if some reason we re-provision or change our infrastructure, someone will need to upload the course, or several version of the course to the new bucket, also this gives more engineering independence to CS team to work, update and version the course, it also assures the course will be always in tar.gz format, has the correct name, correct folder structure, etc.
> I also did a quick search about github releases availability and I didn't find anything that suggest that is less reliable that github itself.

Makes sense. GitHub is reliable, but honestly don't depend on that too much, S3 in comparison is way more reliable by design.

So how about we focus on the original problem I suggested: Downloading from GitHub everytime there's a new customer is slow and error-prone. How about the following solution:

 - During deployment (aka Ansible) time download the course to `/edx/var/edxapp/course-template/course.tar.gz`
 - Extract the course into `/edx/var/edxapp/course-template/course_data`
 - So the task would just import the code

If that sounds too much work, it could justify doing it as a followup work after merging and testing this PR on Staging.

What do you think @melvinsoft?

melvinsoft commented:
@OmarIthawi Thanks for the review and the suggestion. It makes sense, and is more robust, but I won't have time to do that work this sprint, unless I sacrifice other priorities, and this is a Q1 goal. Also, that solution is not compatible with one of the requirements, that is the ability to update the course in place, in that solution we need to re deploy edx platform in order to update the course for new sign ups.

I think the current solution is robust enough, I don't see why we should expect to many errors from Github, also isn't really slow, and it also runs in a celery tasks. 

So we have two options we merged as it is, which I think is robust enough and meets the requirements, or we use a single download URL but it will cost us in engineering time on every course update.

OmarIthawi commented:
Thanks @melvinsoft for sharing the full picture. I'll take another round of review to spot minor issues.

melvinsoft commented:
@thraxil Thanks for your review! 

@OmarIthawi I'm planning to merge this tomorrow so we can start testing on Staging, please give another pass if you can before.

OmarIthawi commented:
@melvinsoft I took the liberty of cleaning up the branch history, please let me know if you'd like me to revert the commit squash.

melvinsoft commented:
@OmarIthawi no worries, thanks for squashing them and thank you both for the reviews.
